lib/util.c

```diff
@@ -2058,3 +2058,32 @@ EXPORTED void tcp_disable_nagle(int fd)
         syslog(LOG_ERR, "unable to setsocketopt(TCP_NODELAY): %m");
     }
 }
+
+EXPORTED void xsyslog_fn(int priority, const char *description,
+                         const char *file_loc, int line_loc,
+                         const char *extra_fmt, ...)
+{
+    static struct buf buf = BUF_INITIALIZER;
+    int saved_errno = errno;
+
+    buf_reset(&buf);
+    buf_appendcstr(&buf, description);
+    buf_appendmap(&buf, ": ", 2);
+    if (extra_fmt && *extra_fmt) {
+        va_list args;
+
+        va_start(args, extra_fmt);
+        buf_vprintf(&buf, extra_fmt, args);
+        va_end(args);
+
+        buf_putc(&buf, ' ');
+    }
+    if (saved_errno) {
+        buf_printf(&buf, "error=<%s> ", strerror(saved_errno));
+    }
+    buf_printf(&buf, "srcfile=<%s> srcline=<%d>", file_loc, line_loc);
```

Comment:
I ran a few callgrind tests and buf_printf for simple `<%s>` string substitutions performs worse by factor 4 than using a combo of buf_putc, buf_appendcstr. If we log lots of stuff in a loop that cost could show up. If we want to omit that cost, we'd need a buf helper function to also encode the line number, though.